contracts/bitizens/AvatarItemToken.sol

```diff
@@ -6,218 +6,280 @@ import "./AvatarItemService.sol";
+    require(_index < burnedTokenIds.length, "out of boundary");
+    return burnedTokenIds[_index];
+  }
+  
+  function getSameItemTokenIds(uint256 _tokenId) external view validToken(_tokenId) returns(uint256[]) {
+    return sameItemTokenIds[_getItemHash(_tokenId)];
   }
 
-  function isBurned(uint256 _tokenId) external view returns (bool) {
-    return isBurnByTokenId[_tokenId];
+  function getItemHash(uint256 _tokenId) external view validToken(_tokenId) returns (bytes32) {
+    return _getItemHash(_tokenId);
   }
 
-  function getBurnTokenCount() external view returns (uint256) {
-    return burnTokenIds.length;
+  function isSameItem(uint256 _tokenId1, uint256 _tokenId2)
+  external
+  view
+  validToken(_tokenId1)
+  validToken(_tokenId2)
+  returns (bool _isSame)
+  {
+    if(_tokenId1 == _tokenId2) {
+      _isSame = true;
+    } else {
+      _isSame = _getItemHash(_tokenId1) == _getItemHash(_tokenId2);
+    }
   }
 
-  function getBurnTokenIdByIndex(uint256 _index) external view returns (uint256) {
-    require(_index < burnTokenIds.length);
-    return burnTokenIds[_index];
+  function addMineTrack(uint256 _tokenId, uint8 _rarity) external onlyOperator {
+    require(exists(_tokenId), "token error");
+    mineTrack[_tokenId].push(_rarity);
   }
 
-  function burnToken(address _owner, uint256 _tokenId) external onlyOperator {
+  function burnToken(address _owner, uint256 _tokenId) external onlyOperator existToken(_tokenId) {
     _burnToken(_owner, _tokenId);
   }
-  
-  function createAvatarItem( 
-    address _owner, 
-    uint32 _listNumber,
-    uint16 _setNumber,
+
+  function createToken( 
+    address _owner,
+    address _founder,
+    address _creator, 
     bool _isBitizenItem, 
-    bool _isSet, 
-    uint8 _types,
+    int16 _probability,
+    uint256[4] _attr1,
+    uint8[5] _attr2) 
+    external  
+    onlyOperator
+    returns(uint256 _tokenId) {
+    require(_owner != address(0), "address invalid");
+    Item memory item = _mintToken(_founder, _creator, _isBitizenItem, _probability, _attr1, _attr2);
+    _tokenId = ++tokenIndex;
+    tokenInfo[_tokenId] = item;
+    _mint(_owner, _tokenId);
+    sameItemTokenIds[_getItemHash(_tokenId)].push(_tokenId);
+  }
+
+  function updateToken(
+    uint256 _tokenId,
+    bool  _isBitizenItem,
+    int16 _probability,
+    uint256 _node,
+    uint256 _listNumber,
+    uint256 _setNumber,
+    uint256 _quality,
     uint8 _rarity,
     uint8 _socket,
     uint8 _gender,
-    uint8 _ext1,
-    uint8 _ext2
-    ) 
-    external 
-    onlyOperator 
-    returns(uint256 _tokenId){
-    _tokenId = _createAvatarItem(_owner, _listNumber, _setNumber, _isBitizenItem, _isSet, _types, _rarity, _socket, _gender, _ext1, _ext2);
+    uint8 _energy,
+    uint8 _ext
+  ) 
+  external 
+  onlyOperator
+  existToken(_tokenId) 
+  {
+    Item storage item = tokenInfo[_tokenId];
+    if (item.isBitizenItem != _isBitizenItem)
```

Comment:
Is it necessary to check? Why not just assign the value in all cases?

Reply:
Now I think it no necessary, haha